Refactor and update to new 1.x API

ahmadia opened the pull request:
- Refactored functionality out of Nutch into JobClient and Job classes
- Added NutchException
- Associate configurations with Nutch
- Associate crawls with JobClient
- Provide filtering of unrelated jobs in JobClient
- Move job shortcuts to JobClient
- Add ConfigClient and Config types
- Add SeedClient and Seed types
- Add tests for above types and basic functionality

Verified functionality:
- Job/JobClient interface
- Config interface
- Seed list injection
- Full-bore crawls

Unverified functionality:
- Admin interface

chrismattmann committed:
Merge pull request #1 from memex-explorer/ahmadia/python3-support

Python 3 support

ahmadia committed:
Refactor and update to new 1.x API

* Refactored functionality out of Nutch into JobClient and Job classes
* Added NutchException
* Associate configurations with Nutch
* Associate crawls with JobClient
* Provide filtering of unrelated jobs in JobClient
* Move job shortcuts to JobClient

Verified functionality:

* Job/JobClient interface

Unverified functionality:

* Admin interface
* Config interface
* Seed list injection
* Full-bore crawls

chrismattmann committed:
Update README.md

- add @ahmadia as a contributor

ahmadia commented:
cc @chrismattmann.  I'll make better progress now that I've got a better handle on this code and its interactions with the server.

chrismattmann commented on nutch/nutch.py:
```diff
@@ -182,63 +149,201 @@ def delete(self, id=None):
+            time.sleep(1)
+            print(self.jobGetInfo(job.id)[0])
+            print('\nPress return to proceed to next step.\n', file=sys.stderr)
+            sys.stdin.read(1)
```
I don't think we should make crawl interactive. It just should print status like in L317, and then move on.

ahmadia replied:
I didn't touch that part of the code.  I agree that it shouldn't be interactive.

chrismattmann commented:
Overall great work. I think you can commit this, but I'd appreciate making crawl not interactive. Can you also issue the same PR to my repo, or I can do that too I'm trying to keep them in sync, but maintain a delineation for my class.

ahmadia commented:
If you don't mind, I'll get a few more commits in so that it's more or less full-featured before I merge in.

ahmadia commented:
Probably by the middle of next week.

chrismattmann commented:
I'm going to go ahead and commit what you have here to my chrismattmann:master. Makes it easier to play with it and test.

chrismattmann committed:
Merge branch 'ahmadia/refactored_api' of https://github.com/memex-explorer/nutch-python into merge-new-api

chrismattmann committed:
- cleanup and use ConfigClient instead of Config

chrismattmann commented:
I made some fixes in https://github.com/memex-explorer/nutch-python/commit/0bc86d2f9c422abf112a398d25263871b376ccb5

ahmadia commented:
Thanks!

ahmadia committed:
More refactoring

* New Server type for low-level interactions
* Better exceptions
* ConfigClient/Config (not quite complete)
* Better default responses

ahmadia commented:
Another refactor, this time I'm adding a low-level Server class to replace the `callServer` function.  I'm also trying to make the interface look more like `elasticsearch-py`, where possible.

chrismattmann commented:
Quick thoughts - why keep refactoring? I mean it was functional right?

ahmadia commented:
It was really hard to work with from Python, and there were a number of things that weren't working due to the changes in the REST API.

chrismattmann commented:
Hmm, well I think these are two orthogonal things. If there were bugs we fix those bugs. But refactoring it to look like ESpy is aesthetic and I'm not getting it and unfortunately not supportive of doing that. I'd rather it look like tika-python and thus callServer I think is something we should keep around, etc. We can talk about this more at the hackathon.

ahmadia commented:
A simple case is the output from listing `/jobs`.  In the current version, that returns a bunch of JSON and a status code.  In the refactored version, a non-200 status raises an exception and the response is converted into a Job object, which in turn has methods associated with it.

If you want to keep `callServer` around, I can provide a similar interface that wraps the Server.call() function (or alternatively, pull callServer back out and have Server.call() wrap it).  My main motivations for providing a Server object is that I want to be able to pass information about how to call Nutch around between the various sub-objects, and I wanted to include bits of context like "raise errors on non-200 status codes" that could be easily set in the top-level Nutch object then forgotten about.

Some of this is aesthetic, and some of it has to do with encapsulating the JSON into usable code.  Let's chat more about this at the hackathon.

ahmadia committed:
Fix a few bugs, add some tests, improve Job/Config

* Implement support for correctly comparing jobs (in/eq/neq)
* Test that jobs can be created, compared, stopped

ahmadia committed:
Add more tests, fix bugs they've identified

* remove namespace package code in __init__.py
* replace CommonEqualityMixin with IdEqualityMixin
* don't use a mutable for a default parameter to Jobs()
* add tests!

ahmadia committed:
Correct a few mistakes in nutch.main()

ahmadia committed:
Implement and test Config interface

ahmadia committed:
Added Seed interactions and tests

Add workaround for https://issues.apache.org/jira/browse/NUTCH-2123

ahmadia commented:
@chrismattmann - Are you still feeling unhappy with this?  I can pull the callServer code back out if it's the only barrier to merging.  Otherwise I can keep working on this branch and we can discuss at the hackathon.  I'm fine with merging in whatever functionality you like and discarding the rest.

ahmadia commented on nutch/__init__.py:
```diff
@@ -14,8 +14,4 @@
-except ImportError:
-    from pkgutil import extend_path
-    __path__ = extend_path(__path__, __name__)
+from .nutch import Nutch, NutchException, Job, Config
```
note, deleting the above is mandatory as soon as we start making stuff accessible in the top-level nutch namespace.  namespace packages are a real PITA so it's better to just get rid of this now.

ahmadia commented on nutch/nutch.py:
```diff
@@ -56,189 +56,404 @@
+            else:
+                warn('Nutch server returned status:', resp.status_code)
+        content_type = resp.headers['content-type']
+        if content_type == 'application/json' and not forceText:
```
forceText is the workaround for the response not matching the headers (for now).

ahmadia commented on nutch/nutch.py:
```diff
@@ -56,189 +56,404 @@
+        new_config = Config(cid, self.server)
+        return new_config
+
+    def __getitem__(self, item):
```
this provides nc.Configs()['default']

ahmadia commented on nutch/nutch.py:
```diff
@@ -56,189 +56,404 @@
+        # not found!
+        raise KeyError
+
+    def __setitem__(self, key, value):
```
I couldn't get this working

ahmadia commented on nutch/nutch.py:
```diff
@@ -56,189 +56,404 @@
         command = command.upper()
         if command not in LegalJobs:
-            warn('Nutch command must be one of: %s' % LegalJobs.join(', '))
+            warn('Nutch command must be one of: %s' % ', '.join(LegalJobs))
```
I think this was a bug.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+# encoding: utf-8
```
TESTS!

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+
+# Test Nutch API
+# Assumes a Nutch REST server is running on localhost
+# TODO: Package into Travis tests
```
With Travis, we can run PRs against a Travis configuration with a Nutch REST server installed.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+import nutch
+import pytest
+import glob
+
```
All of the tests are interacting with the REST server, so these are more integration tests than unit tests.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+    # there has to be something smarter to check here
+    assert default_config.info()
+
+# I don't know how to get this working
```
This is the functionality I can't figure out how to get working.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+    assert(len(updated_jobs) == len(old_jobs) + 1)
+
+    # awesome functionality for checking if this job is in a list of jobs
+    assert(inject_job not in old_jobs)
```
I think this is pretty sweet.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+
+    jc1_job = jc1.inject()
+
+    # only jobs with the same crawlId are returned in the list()
```
This is pretty sweet too.

ahmadia commented on nutch/test_nutch.py:
```diff
@@ -0,0 +1,140 @@
+    inject_job.abort()
+    assert(inject_job.info()['state'] == 'KILLED')
+
+# How do we delete jobs using the REST API?  Is it even possible?
```
Is a delete method exposed?  I didn't see it anywhere in the documentation.

ahmadia committed:
Improve inject, fix default crawl id, better tests

* Nutch doesn't like ':' in paths, remove from default crawl id
* Add Queue class.  This will eventually be used to properly execute a
  crawl cycle, and to ensure that a previous job has completed
  successfuly before launching the next one.
* improve the tests

ahmadia committed:
Basic Crawl object

ahmadia committed:
full crawl cycle and tests

ahmadia committed:
Working Nutch crawls from Python

* Add Googlebot as default user string if not specified
* improved tests

ahmadia commented:
@chrismattmann - this is ready for review/merge.

chrismattmann commented:
I love this. I am +1 to merge, and thus am merging it. You are awesome @ahmadia thank you.
